src/sinks/util/socket_events_counter.rs

```diff
@@ -0,0 +1,169 @@
+use crate::{buffers::Acker, Event};
+use bytes::Bytes;
+use futures::{ready, Sink};
+use pin_project::{pin_project, pinned_drop};
+use std::{
+    io::{Error as IoError, ErrorKind},
+    marker::Unpin,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll},
+};
+use tokio::io::AsyncWrite;
+use tokio_util::codec::{BytesCodec, FramedWrite};
+
+const MAX_PENDING_ITEMS: usize = 10_000;
+
+/// Count number of encoded events and ack on request.
+pub struct EventsCounter {
+    sizes: Mutex<Vec<usize>>,
```

Comment:
If we really need to use `Mutex`, let's use `tokio::sync::Mutex` instead to avoid blocking on `lock()`.

Reply:
That's completely okay to use `std::sync::Mutex` in async code since our lock almost instantly. This even mentioned in `tokio::sync::Mutex` docs (https://docs.rs/tokio/0.2.21/tokio/sync/struct.Mutex.html):
> There are some situations where you should prefer the mutex from the standard library. Generally this is the case if:
> The duration of any single lock is near-instant.

I'm not completely sure but I think it's even possible to use `RefCell` here because as I understand there should not be concurrent access to `sizes` from different places. From the other side `Mutex` looks safer and I not sure about the performance difference between `Mutex` and `RefCell`.

Reply:
> That's completely okay to use std::sync::Mutex in async code since our lock almost instantly.
Seem fair.

`AtomicUsize` is also an option.

Reply:
What's the need for `Vec`? Can't we count events and sum their sizes instead? If it's for internal events only, then they can be changed to accept aggregated data.

Reply:
We can drop `Vec` in favour of `AtomicUsize` but only if can submit more events/bytes to internal metrics that actually happen, otherwise we need `Vec` for proper values =\

Reply:
Regarding `Vec`
> if can submit more events/bytes to internal metrics that actually happen

true, I missed that. `Vec` is fine.